src/shady-css/parser.js

```diff
@@ -0,0 +1,229 @@
+          name += tokenCursor.takeOne().text;
+        }
+      } else if (tokenCursor.next.text === '{') {
+        block = this.parseBlock(tokenCursor);
+        break;
+      } else if (matcher.propertyDelimiter.test(tokenCursor.next.text)) {
+        // TODO: why not take one?
+        this.traverseDelimiter(tokenCursor);
+        break;
+      } else {
+        if (whitespace && parameters) {
+          parameters += whitespace;
+        }
+        whitespace = '';
+        parameters += tokenCursor.takeOne().text;
+      }
+    }
+
+    return this.nodeFactory.atRule(name, parameters, block);
+  }
+
+  parseBlock(tokenCursor) {
+    let declarations = [];
+
+    // Take the opening { boundary:
+    tokenCursor.takeOne();
+
+    while (tokenCursor.next) {
+      if (tokenCursor.next.type === tokenType.whitespace) {
+        this.traverseWhitespace(tokenCursor);
+      } else if (tokenCursor.next.text === '}') {
+        // Take the closing } boundary:
+        tokenCursor.takeOne();
+        break;
+      } else {
+        let declaration = this.parseDeclaration(tokenCursor);
+        if (declaration == null) {
+          break;
+        }
+        declarations.push(declaration);
+      }
+    }
+
+    return this.nodeFactory.block(declarations);
+  }
+
+  parseRule(tokenCursor) {
+    let rule = '';
+    let whitespace = '';
+
+    while (tokenCursor.next) {
+      if (tokenCursor.next.type === tokenType.whitespace) {
+        whitespace = this.traverseWhitespace(tokenCursor);
+      } else if (tokenCursor.next.text === '(') {
+        rule += this.traverseCall(tokenCursor);
+      } else if (matcher.blockBoundary.test(tokenCursor.next.text)) {
+        break;
+      } else {
+        if (whitespace && rule) {
+          rule += whitespace;
+        }
+        whitespace = '';
+        rule += tokenCursor.takeOne().text;
+      }
+    }
+
+    // A selector never contains or ends with a semi-colon.
+    if (matcher.propertyDelimiter.test(tokenCursor.next.text)) {
+      let colonIndex = rule.indexOf(':');
+      let value = rule.substr(colonIndex + 1).replace(/^\s*/, '');
+
+      this.traverseDelimiter(tokenCursor);
+
+      return this.nodeFactory.property(
+          rule.substr(0, colonIndex),
+          this.nodeFactory.propertyValue(value));
+    } else if (matcher.propertyNameTrail.test(rule)) {
+      let block = this.parseBlock(tokenCursor);
+
+      this.traverseDelimiter(tokenCursor);
+
+      return this.nodeFactory.property(rule.substr(0, rule.length - 1), block);
+    } else {
+      return this.nodeFactory.selector(rule, this.parseBlock(tokenCursor));
+    }
+  }
+
+  traverseCall(tokenCursor) {
+    let call = '';
+
+    if (tokenCursor.next.text === '(') {
+      while (tokenCursor.next.text !== ')') {
+        call += tokenCursor.takeOne().text;
+      }
+    }
+
+    return call;
+  }
+
+  traverseWhitespace(tokenCursor) {
+    let whitespace = '';
+
+    while (tokenCursor.next &&
```

Comment:
This was done in the hopes of achieving a performance boost. I figured that it was an acceptable optimization to treat consecutive whitespace as one whole. This is also done for a few other classes of characters.